packages/react/src/components/DateTimePicker/dateTimePickerUtils.js

```diff
@@ -794,3 +794,102 @@ export const useCloseDropdown = ({
     // eslint-disable-next-line react-hooks/exhaustive-deps
     [defaultValue, isExpanded, setCustomRangeKind, setIsExpanded, lastAppliedValue]
   );
+
+/**
+ * A hook handling the height of the drop down menu
+ *
+ * @param {object} containerRef the ref to the container div of the drop down menu
+ * @param {boolean} isSingleSelect if it is single select calendar
+ * @param {boolean} isCustomRange if dropdown was opened in custom range
+ * @param {boolean} showRelativeOption are the relative options shown by default
+ * @param {string} customRangeKind custom range kind is either relative or absolute
+ * @param {function} setLeft set scrollLeft
+ * @param {function} setTop set scrollTop
+ * @returns Object An object containing:
+ *    offTop (boolean): if the menu is off top
+ *    offBottom (boolean): if the menu is off bottom
+ *    inputTop (string) : the top position of the date time input
+ *    inputBottom (string): the bottom position of the date time input
+ *    customHeight (string): the adjusted height of the drop down menu if both offTop and offBottom are true
+ *    maxHeight (string) : maximum height of the drop down menu
+ */
+export const useCustomHeight = ({
+  containerRef,
+  isSingleSelect,
+  isCustomRange,
+  showRelativeOption,
+  customRangeKind,
+  setScrollLeft,
+  setScrollTop,
+}) => {
+  // calculate max height for varies dropdown
+  const presetMaxHeight = 315;
+  const relativeMaxHeight = 269;
+  const withoutRelativeOptionMaxHeight = 446;
+  const absoluteMaxHeight = 555;
+  const singleMaxHeight = 442;
+  const footerHeight = 40;
+  const maxHeight = isCustomRange
+    ? showRelativeOption
+      ? isSingleSelect
+        ? singleMaxHeight
+        : customRangeKind === PICKER_KINDS.ABSOLUTE
+        ? absoluteMaxHeight
+        : relativeMaxHeight
+      : withoutRelativeOptionMaxHeight
+    : presetMaxHeight;
+
+  // get new scroll top and scroll left when scrolling the input
+  const getPosition = (event) => {
+    setScrollTop(event.target.scrollTop);
+    setScrollLeft(event.target.scrollLeft);
+  };
+
+  const debouncedPosition = debounce(getPosition, 10, {
+    leading: false,
+    trailing: true,
+  });
+
+  // add event listener on the scroll events of the parents
+  useEffect(() => {
+    let currentNode = containerRef?.current?.parentNode;
+    const parentNodes = [];
+    while (currentNode) {
+      parentNodes.push(currentNode);
+      currentNode.addEventListener('scroll', debouncedPosition);
+      currentNode = currentNode.parentNode;
+    }
```

Comment:
I think this could get expensive. If this component is deeply nested in the DOM this looks to put a scroll listener on every element in that parent chain. Even with the debounce. I worry it could have performance issues. @sls-ca thoughts?